class MinStack {
    Stack<Long> st = new Stack<>();
    Long min;
    public MinStack() {
        min = Long.MAX_VALUE;
    }
    
    public void push(int val) {
        Long value = Long.valueOf(val);

        if(st.isEmpty()){
            st.push(value);
            min = value;
        }
        else{
            if(value >= min){
                st.push(value);
            }
            else{
                st.push(2 * value - min);
                min = value;
            }
        }
    }
    
    public void pop() {
        if(st.isEmpty()){
            return;
        }

        Long val = st.pop();
        if(val < min){
            min = 2 * min - val;
        }
        
    }
    
    public int top() {
        Long val = st.peek();

        if(val < min){
            return min.intValue();
        }

        return val.intValue();
    }
    
    public int getMin() {
        return min.intValue();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(value);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */